### PR TITLE
more random functions

### PR DESCRIPTION
**Pull request by catoverdrive**
(mostly the ones that balding nichols uses.)

### CODE REVIEW

File: python/hail/expr/functions.py [hunk 1]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+@typecheck(a=expr_float64,
+           scale=expr_float64,
+           seed=nullable(int))
+def rand_gamma(a, scale, seed=None) -> Float64Expression:
+    """Returns a random floating-point number drawn from a gamma distribution
+    with parameters `a` and `scale`.
+
+    Notes
+    -----
+    The optional parameters `min` and `max` represent a truncated beta
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: delete paragraph

File: python/hail/expr/functions.py [hunk 2]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+
+    >>> hl.rand_gamma(1, 1).value
+    0.1339939936379711
+
+    Parameters
+    ----------
+    a : float or :class:`.Expression` of type :py:data:`.tfloat64`
+    scale : float or :class:`.Expression` of type :py:data:`.tfloat64`
+    seed : :obj:`int` or `None`
+        If not `None`, function will be seeded with provided seed.
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: Random seed.

File: python/hail/expr/functions.py [hunk 3]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+    -------
+    :class:`.Expression` of type :py:data:`.tfloat64`
+    """
+    return _seeded_func("rand_gamma", tfloat64, seed, a, scale)
+
+
+@typecheck(probabilities=expr_array(expr_float64),
+           seed=nullable(int))
+def rand_draw(probabilities, seed=None) -> Int32Expression:
+    """Returns a random index from an array of unnormalized probabilities.
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: Returns a random draw from a categorial distribution. Link to https://en.wikipedia.org/wiki/Categorical_distribution

File: python/hail/expr/functions.py [hunk 4]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+
```
Comment (jbloom22) [inline python/hail/expr/functions.py]:
```
Notes
-----
The categories correspond to the indices of `prob`, an unnormalized probability mass function. The probability of drawing index ``i`` is ``prob[i]/sum(prob)``.
``

File: python/hail/expr/functions.py [hunk 5]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+    Returns
+    -------
+    :class:`.Expression` of type :py:data:`.tfloat64`
+    """
+    return _seeded_func("rand_gamma", tfloat64, seed, a, scale)
+
+
+@typecheck(probabilities=expr_array(expr_float64),
+           seed=nullable(int))
+def rand_draw(probabilities, seed=None) -> Int32Expression:
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: I'd rename this to `rand_cat` and the variable to `prob`, e.g. https://www.rdocumentation.org/packages/extraDistr/versions/1.8.9/topics/Categorical

File: python/hail/expr/functions.py [hunk 6]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+    -------
+    :class:`.Expression` of type :py:data:`.tfloat64`
+    """
+    return _seeded_func("rand_draw", tint32, seed, probabilities)
+
+
+@typecheck(a=expr_array(expr_float64),
+           seed=nullable(int))
+def rand_dirichlet(a, seed=None) -> ArrayExpression:
+    """Returns a sample from the Dirichlet distribution parameterized by vector
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: "Returns a random draw from a Dirichlet distribution." with link to 
https://en.wikipedia.org/wiki/Dirichlet_distribution

File: python/hail/expr/functions.py [hunk 7]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+        max = hl.literal(1)
+
+    return _seeded_func("rand_beta", tfloat64, seed, a, b, min, max)
+
+
+@typecheck(a=expr_float64,
+           scale=expr_float64,
+           seed=nullable(int))
+def rand_gamma(a, scale, seed=None) -> Float64Expression:
+    """Returns a random floating-point number drawn from a gamma distribution
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: "Returns a random draw from a gamma distribution." and link https://en.wikipedia.org/wiki/Gamma_distribution

File: python/hail/expr/functions.py [hunk 8]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+    if max is None:
+        max = hl.literal(1)
+
+    return _seeded_func("rand_beta", tfloat64, seed, a, b, min, max)
+
+
+@typecheck(a=expr_float64,
+           scale=expr_float64,
+           seed=nullable(int))
+def rand_gamma(a, scale, seed=None) -> Float64Expression:
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: I think it will be clearer to rename `a` to `shape`.

https://en.wikipedia.org/wiki/Gamma_distribution

Comment (jbloom22) [reply python/hail/expr/functions.py]: And make sure scale is really scale and not rate

File: python/hail/expr/functions.py [hunk 9]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+
+    >>> hl.rand_dirichlet([1, 1, 1]).value
+    [0.4630197581640282,0.18207753442497876,0.3549027074109931]
+
+    >>> hl.rand_dirichlet([1, 1, 1]).value
+    [0.20851948405364765,0.7873859423649898,0.004094573581362475]
+
+    Parameters
+    ----------
+    a : :obj:`list` of float or :class:`.ArrayExpression` of type :py:data:`.tfloat64`
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: add description:
Array of non-negative concentration parameters.

File: python/hail/expr/functions.py [hunk 10]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+    seed : :obj:`int` or `None`
+        If not `None`, function will be seeded with provided seed.
+
+    Returns
+    -------
+    :class:`.Expression` of type :py:data:`.tfloat64`
+    """
+    return hl.bind(lambda x: x / hl.sum(x),
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: slick

File: python/hail/methods/statgen.py [hunk 11]
```diff
@@ -2059,8 +2059,7 @@ def balding_nichols_model(n_populations, n_samples, n_variants, n_partitions=Non
     >>> bn_ds = hl.balding_nichols_model(4, 40, 150, 3,
     ...          pop_dist=[0.1, 0.2, 0.3, 0.4],
     ...          fst=[.02, .06, .04, .12],
-    ...          af_dist=TruncatedBetaDist(a=0.01, b=2.0, min=0.05, max=1.0),
-    ...          seed=1)
+    ...          af_dist=hl.rand_beta(a=0.01, b=2.0, min=0.05, max=1.0))
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: how to seed?

File: python/hail/methods/statgen.py [hunk 12]
```diff
@@ -2207,6 +2204,7 @@ def balding_nichols_model(n_populations, n_samples, n_variants, n_partitions=Non
                                             mixture)
     return MatrixTable(jmt)
 
+
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: remove line

File: src/main/scala/is/hail/expr/ir/functions/RandomSeededFunctions.scala [hunk 13]
```diff
@@ -28,28 +29,44 @@ class IRRandomness(seed: Long) {
   def rpois(lambda: Double): Double = Poisson.random(lambda, random, poisState)
 
   def rnorm(mean: Double, sd: Double): Double = mean + sd * random.nextGaussian()
+
+  def rbeta(a: Double, b: Double): Double = Beta.random(a, b, random)
+
+  def rgamma(a: Double, scale: Double): Double = Gamma.random(a, scale, random)
+
+  def rdraw(a: Array[Double]): Int = {
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/functions/RandomSeededFunctions.scala]: `rcat(prob)`

File: src/main/scala/is/hail/expr/ir/functions/RandomSeededFunctions.scala [hunk 14]
```diff
@@ -28,28 +29,44 @@ class IRRandomness(seed: Long) {
...
+    val sum = a.sum
+    val probs = Array.tabulate(a.length)(i => a(i) / sum)
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/functions/RandomSeededFunctions.scala]: no need for rescaling. Just set `draw = random.nextDouble() * sum`

File: python/hail/expr/functions.py [hunk 15]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+
+
+@typecheck(probabilities=expr_array(expr_float64),
+           seed=nullable(int))
+def rand_draw(probabilities, seed=None) -> Int32Expression:
+    """Returns a random index from an array of unnormalized probabilities.
+
+    Warning
+    -------
+    This function may be slow over very large arrays.
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: This function may be slow when the number of categories is large.

File: python/hail/expr/functions.py [hunk 16]
```diff
@@ -1778,7 +1778,159 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+@typecheck(a=expr_array(expr_float64),
+           seed=nullable(int))
+def rand_dirichlet(a, seed=None) -> ArrayExpression:
+    """Returns a sample from the Dirichlet distribution parameterized by vector
+    `a`.
+
+    Examples
+    --------
+
+    >>> hl.rand_dirichlet([1, 1, 1]).value
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: what behavior do you get if a value is negative?

File: python/hail/methods/statgen.py (hunk 11 above)
Comment (catoverdrive) [reply python/hail/methods/statgen.py]: ugh, somehow some changes from unrelated commits snuck in here. should be gone now.

File: python/hail/expr/functions.py (hunk 16 above)
Comment (catoverdrive) [reply python/hail/expr/functions.py]:
```
>>> hl.rand_dirichlet([-1, -1, 1]).value
[nan, nan, nan]
```

File: python/hail/expr/functions.py [hunk 17]
```diff
@@ -1710,23 +1711,23 @@ def rand_norm(mean=0, sd=1, seed=None) -> Float64Expression:
...
-    :class:`.Expression` of type :py:data:`.tfloat64`
+    :class:`.Float64Expression`
     """
     return _seeded_func("rand_norm", tfloat64, seed, mean, sd)
 
 
 @typecheck(lamb=expr_float64, seed=nullable(int))
 def rand_pois(lamb, seed=None) -> Float64Expression:
-    """Samples from a Poisson distribution with rate parameter `lamb` (RNG).
+    """Samples from a Poisson distribution with rate parameter `lamb`.
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: please change parameter name to lambda and link to https://en.wikipedia.org/wiki/Poisson_distribution

File: python/hail/expr/functions.py [hunk 18]
```diff
@@ -1858,54 +1853,61 @@ def rand_gamma(a, scale, seed=None) -> Float64Expression:
...
 
 
-@typecheck(probabilities=expr_array(expr_float64),
+@typecheck(prob=expr_array(expr_float64),
            seed=nullable(int))
-def rand_draw(probabilities, seed=None) -> Int32Expression:
-    """Returns a random index from an array of unnormalized probabilities.
+def rand_cat(prob, seed=None) -> Int32Expression:
+    """Samples from a `categorical distribution
+    <https://en.wikipedia.org/wiki/Categorical_distribution>`__
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: add period

File: python/hail/expr/functions.py [hunk 19]
```diff
@@ -1767,36 +1768,37 @@ def rand_unif(min, max, seed=None) -> Float64Expression:
...
+           lower=nullable(expr_float64),
+           upper=nullable(expr_float64),
            seed=nullable(int))
-def rand_beta(a, b, min=None, max=None, seed=None) -> Float64Expression:
-    """Returns a random floating-point number drawn from a beta distribution
-    with parameters a and b.
+def rand_beta(a, b, lower=None, upper=None, seed=None) -> Float64Expression:
+    """Samples from a `beta distribution
+    <https://en.wikipedia.org/wiki/Beta_distribution>`__ with parameters `a`
+    (alpha) and `b` (beta).
```
Comment (jbloom22) [inline python/hail/expr/functions.py]: nice, very clear

File: python/hail/expr/functions.py (hunk 17 above)
Comment (catoverdrive) [reply python/hail/expr/functions.py]: `lambda` is a reserved word in python, unfortunately

**Comment by jbloom22** [comment]
don't forget to make a discuss post on the breaking change on the parameter names (lower/upper)

**Pull request merged**